File: scratch/repair_encoding.js
const fs = require('fs');
const path = require('path');

// We need a way to encode/decode windows-1252 in Node.
// Since we don't want to install extra packages, we can use a simple mapping 
// or just trust that for most common mojibake, the byte values match Latin-1 
// except for the 0x80-0x9F range.
// Fortunately, most mojibake characters are in the 0xC0-0xFF range which match.

function fixMojibake(content) {
    let current = content;
    while (true) {
        try {
            // Convert string to bytes using Latin-1 (preserves byte values 0-255)
            const bytes = Buffer.from(current, 'latin1');
            
            // Try to decode those bytes as UTF-8
            // If it contains invalid UTF-8, this might not throw but will produce replacement chars.
            // We want to detect if it's "valid" UTF-8.
            const decoded = bytes.toString('utf8');
            
            // If decoding produced replacement characters or didn't change anything, stop.
            if (decoded === current || decoded.includes('\ufffd')) {
                break;
            }
            
            // Heuristic: if the new string is longer or has more 'weird' chars, it might be wrong.
            // But usually mojibake strings are longer than the original.
            // Let's just check if it's stable.
            current = decoded;
        } catch (e) {
            break;
        }
    }
    return current;
}

const files = [
    'js/admin.js',
    'js/checklist.js',
    'js/dashboard-ui.js',
    'app.js',
    'registro.html',
    'dashboard.html',
    'index.html'
];

const workspaceRoot = 'c:\\Users\\holamundo\\Documents\\GitHub\\adestacion';

files.forEach(f => {
    const filePath = path.join(workspaceRoot, f);
    if (fs.existsSync(filePath)) {
        console.log(`Processing ${f}...`);
        const content = fs.readFileSync(filePath, 'utf8');
        const fixed = fixMojibake(content);
        if (fixed !== content) {
            console.log(`  Fixed mojibake in ${f}`);
            fs.writeFileSync(filePath, fixed, 'utf8');
        } else {
            console.log(`  No mojibake found in ${f}`);
        }
    } else {
        console.log(`File not found: ${f}`);
    }
});

File: scratch/repair_encoding.py
import os
import re

def fix_mojibake(text):
    """
    Attempts to fix mojibake by repeatedly re-encoding as cp1252 and decoding as utf-8.
    """
    current = text
    while True:
        try:
            # Try to 'undo' one level of encoding mismatch
            # We use windows-1252 (cp1252) because it's the most common source of 
            # characters like '€' or '—' being misinterpreted in Windows environments.
            candidate = current.encode('cp1252').decode('utf-8')
            if candidate == current:
                break
            # If the candidate has FEWER non-ascii characters than current, it's likely a fix.
            # Or if it just succeeds, we take it as a potential improvement.
            current = candidate
        except (UnicodeEncodeError, UnicodeDecodeError):
            break
    return current

def process_file(filepath):
    print(f"Processing {filepath}...")
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()
    except UnicodeDecodeError:
        print(f"  Error: Could not read {filepath} as UTF-8. Trying Latin-1...")
        with open(filepath, 'r', encoding='latin-1') as f:
            content = f.read()
    
    fixed_content = fix_mojibake(content)
    
    if fixed_content != content:
        print(f"  Fixed mojibake in {filepath}")
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(fixed_content)
    else:
        print(f"  No mojibake found in {filepath}")

files_to_fix = [
    'js/admin.js',
    'js/checklist.js',
    'js/dashboard-ui.js',
    'app.js',
    'registro.html',
    'dashboard.html',
    'index.html'
]

# Get absolute paths
workspace_root = r'c:\Users\holamundo\Documents\GitHub\adestacion'
for rel_path in files_to_fix:
    abs_path = os.path.join(workspace_root, rel_path.replace('/', os.sep))
    if os.path.exists(abs_path):
        process_file(abs_path)
    else:
        print(f"File not found: {abs_path}")
